Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py

```diff
@@ -0,0 +1,693 @@
+                general_context['SSH'][key] = list(self.ssh_data[key])
+        if 'TELNET' in self.extracted_protocols:
+            general_context['Telnet'] = {'Commands': list(self.telnet_commands),
+                                         'Data': list(self.telnet_data)}
+        if is_flows:
+            general_context['Flow'] = flows_to_ec(self.flows)
+        if is_reg_extract:
+            general_context['IP'] = list(self.ips_extracted)
+            general_context['URL'] = list(self.urls_extracted)
+            general_context['Email'] = list(self.emails_extracted)
+        if self.homemade_regex:
+            general_context['Regex'] = list(self.homemade_extracted)
+        if self.unique_ips:
+            all_ips = self.unique_source_ip.copy()
+            all_ips.update(self.unique_dest_ip)
+            if general_context.get('IP'):
+                general_context.get('IP').append(list(all_ips))  # type: ignore
+            else:
+                general_context['IP'] = list(all_ips)
+
+            general_context['SourceIP'] = list(self.unique_source_ip)
+            general_context['DestIP'] = list(self.unique_dest_ip)
+        ec = {'PcapResults(val.EntryID == obj.EntryID)': general_context}
+        return md, ec, general_context
+
+    def mine(self, file_path: str, decrypt_key: str, is_flows: bool, is_reg_extract: bool, pcap_filter: str,
+             pcap_filter_new_file_path: str):
+        """
+        The main function of the script. Mines the PCAP.
+
+        Args:
+            file_path: The PCAP's file path.
+            decrypt_key: The decryption key (if needed) to decrypt the PCAP.
+            is_flows: Whether to extract flows.
+            is_reg_extract: Whether to extract regexes from the PCAP.
+            pcap_filter: A filter to apply on the PCAP. Same filter syntax as in Wireshark
+            pcap_filter_new_file_path: The new path to save the filtered PCAP in
+
+        """
+
+        try:
+            cap = pyshark.FileCapture(file_path, display_filter=pcap_filter, output_file=pcap_filter_new_file_path,
+                                      decryption_key=decrypt_key, encryption_type='WPA-PWD', keep_packets=False)
+
+            for packet in cap:
+                self.last_layer.add(packet.layers[-1].layer_name)
+
+                layers = str(packet.layers)
+                # remove duplicate layer names such as [ETH,DATA,DATA] -> # [ETH, DATA]
+                layers = list(dict.fromkeys(layers.split(',')))  # type: ignore
+                layers = strip(str(layers))
+                self.hierarchy[layers] = self.hierarchy.get(layers, 0) + 1  # type: ignore
+
+                # update times
+                packet_epoch_time = float(packet.frame_info.get('time_epoch'))
+                self.max_time = max(self.max_time, packet_epoch_time)
+                self.min_time = min(self.min_time, packet_epoch_time)
+
+                # count packets
+                self.num_of_packets += 1
```

Comment:
can't this be outside the loop?
`self.num_of_packets = len(cap)`

Reply:
Not really. As I don't want to load all the pcap to python's memory, the pyshark library doesn't keep the packets - everytime I call for a packet it goes and fetches it. This is what makes this script a bit slow but keeps the memory pretty clean. I think I found a way to load a bunch of packets at once so I'll dig a bit into that maybe to make this faster.